Fix tooltip animation when target changes while animating

jcopperfield opened the pull request:
tooltip in 'index' mode doesn't animate smoothly.

Fixes #4989

simonbrunel commented:
@jcopperfield do you have a fiddle that shows the new behavior ?

jcopperfield commented:
@simonbrunel [codepen example](https://codepen.io/anon/pen/ooQRve) of the new behavior.

simonbrunel commented on src/core/core.tooltip.js:
```diff
@@ -852,27 +852,27 @@ module.exports = function(Chart) {
 
+			// always call me.pivot before returning to
+			// reset me._start for smooth animations issue #4989
+			me.pivot();
```
I don't think `me.pivot()` should be called if the `_active` target(s) didn't change, since it resets the animation start/end values. I think that's what caused the weird animation accelerations in the following gif.

![chartjs 5005](https://user-images.githubusercontent.com/3874900/33437747-2a37cc60-d5e9-11e7-8ae3-35670a99fac2.gif)

jcopperfield replied:
@simonbrunel not calling `me.pivot()` [codepen](https://codepen.io/anon/pen/rYKQrx) causes a much weirder experience by resetting all settings to their initial values. (sorry don't have any software to make a fancy gif).
Any suggestions?

simonbrunel replied:
I can't debug right now :\ Maybe `_active` is different depending on the hovered bar, so the first `changed` value is `true` because of the array order?! Though, I'm not sure why checking the latest value of `changed` doesn't work:

```javascript
  // {...}
  if (changed) {
    me.pivot();
  }

  return changed;
}
```

BTW, the following looks weird (line 869):

```javascript
if (changed) {
    // {...}
    // See if our tooltip position changed
    changed |= (model.x !== me._model.x) || (model.y !== me._model.y);
}
```

because if `changed` is `true` then `changed |= true or false` is necessary `true`, right?

jcopperfield replied:
@simonbrunel I agree that line 869 is obsolete, however it was there already obsolete from a previous pull.
The problem doesn't lay in this handleEvent, but in its caller, which `|=` its `changed` with the result of the `tooltip.handleEvent` (`changed |= tooltip && tooltip.handleEvent(e);`). When the caller has changed, but the tooltip hasn't, the bug presents itself, because the animation will run. I don't understand the animation system well enough to prevent the tooltip animation restarting.

simonbrunel replied:
I'm wondering if it should not be `changed &= (model.x !== me._model.x) || (model.y !== me._model.y);` instead?

simonbrunel replied:
The issue seems to be that we detect a change for the [highlighted item](https://codepen.io/anon/pen/GOwVwQ) resetting the current animation. Calling `pivot()` as you do makes things better in this case but breaks animations when hover interaction mode/intersect are the same as tooltip ([master](https://codepen.io/anon/pen/NwEQme) - [PR](https://codepen.io/anon/pen/zPMgVq)).

jcopperfield commented:
@simonbrunel The latest push is in the handler that is causing the problem. I'm not sure if the animation behaviour with `hoverBackgroundColor` is as it should be, as I have no clue what the preferred behaviour is. Maybe the problem is that the `tooltip` doesn't set the `tooltip.animation` or that different animations are handled by the same process.

jcopperfield commented:
@simonbrunel The new PR seems to be working a bit better than the previous one.
[codepen multiple charts v.2.7 & PR](https://codepen.io/anon/pen/jadxvK)
```javascript
// v2.7
changed |= tooltip && tooltip.handleEvent(e);

// PR
if (tooltip) {
    changed = tooltip.handleEvent(e);
}
```

simonbrunel commented:
It doesn't work, hover events are broken (skipped) as soon as the tooltip reaches its last position (and so returns false to `changed`). It happens when `intersects` and/or `mode` are different between `tooltips` and `hover`.

For example: https://codepen.io/anon/pen/mqvKJB
```javascript
        tooltips: {
            mode: 'index',
            intersect: false
        },
        hover: {
            mode: 'index',
            intersect: true
        },
```

doesn't work if you move the mouse at the same index but over and outside a bar.

jcopperfield commented:
@simonbrunel I see. So should there be a flag `tooltip.animating` implemented in some part to resolve this problem?

simonbrunel commented:
If think the problem is that animations are handled globally and thus are not independent from each others, so when there is a change anywhere, we restart all current animations (I'm not even sure we can have animations with different durations?!) Ideally, animations should be handled separately and locally: the tooltip plugin should be able to trigger an animation with its own parameters, stop/restart this animation without impacting others, etc.

But I think that would require a rewrite of the animation service which is out the scope of this PR (and could be a bit complicated). Though, I'm investigating this solution to animate stuff in the datalabels plugin. Right now I don't think of a correct way to easily fix that issue, the flag suggestion could "maybe" work but will certainly be hacky to synchronize :\

jcopperfield commented:
@simonbrunel I agree that a per element animation service would be the optimal solution.

Introducing a `setAnimating` argument in `transition` to flag the tooltip as `animating` allows for a finer grained checking to resolve the case:
```javascript
tooltips: {
    mode: 'index',
    intersect: false
},
hover: {
    mode: 'index',
    intersect: true
}
```
[example PR](https://codepen.io/anon/pen/gXqjGV)

simonbrunel commented on src/core/core.element.js:
```diff
@@ -35,12 +35,10 @@ function interpolate(start, view, model, ease) {
-						view[key] = c1.mix(c0, ease).rgbString();
-						continue;
-					}
+				c1 = color(target);
```
Why changing that part of the code? Your changes actually decrease performances because `c0` and `c1` are now always evaluated (`color()`), which is a very costly operation while transitioning values.

simonbrunel commented:
I would avoid adding a flag to the base element class to workaround an issue with the animation service and especially with the tooltip. It's also semantically incorrect since you force the element to report an animating state while it's not when `!model || ease === 1` (very confusing).

If you really want to rely on a flag, I would keep that workaround at the highest level, so in the tooltip class and private `_skipUpdateWorkaround` to avoid anyone from using it (it's very tempting to use `element.animating` but it doesn't work as expected!)

Don't take care of the newly reported "cognitive complexity" of CodeClimate, I'm going to increase it.

jcopperfield commented:
@simonbrunel the `me.animating` flag only applies to the tooltip, not to the `element.animating`. Unfortunately the `tooltip.transition` code is called from inside the `core.element.js`. When `(!model || ease === 1)` the `me.animating` flag is removed, because the animation/transition has finished. Don't really understand what the confusing part is. Could you explain?
I could rename the `setAnimating` to `isTooltip` if that would makes things clearer.

simonbrunel commented:
The reviewed code was [this one](/chartjs/Chart.js/pull/5005/commits/38574c38fec3905949a66824785e635e2811f8dd#diff-1b6ffa7bdff6275f06e48ba2af033e99), I didn't notice you changed it for `delete me.animating` in your last commit.

jcopperfield commented:
I will rename the `setAnimating` to `isTooltip` and `me.animating` to `me._skipUpdateWorkaround` to prevent misusing the flag.

simonbrunel commented:
What about using (still as dirty workaround) the `element._start` state: if `!null`, that's mean it's animating:

```javascript
changed = tooltip._start     // === animating
  ? tooltip.handleEvent(e)
  : changed | tooltip.handleEvent(e);
```

That would allow to not touch the element class.

jcopperfield commented:
That might work. I will try that.

jcopperfield commented:
[example of last PR using tooltip._start](https://codepen.io/anon/pen/mqvvVB)

jcopperfield commented:
@simonbrunel Is there anything missing from this PR or PR 4959 that prevents merging?
